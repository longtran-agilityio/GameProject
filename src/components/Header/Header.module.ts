// lib
import { Box, Typography, styled } from '@mui/material'
import { Link } from 'react-router-dom'

// theme
import { theme } from '@webapp/themes/index'

export const StyledHeader = styled('header')(({ theme }) => ({
  position: 'fixed',
  top: 0,
  left: 0,
  right: 0,
  zIndex: 1,
  backgroundColor: theme.palette.background.default,
  padding: theme.spacing(2, 5),
  boxSizing: 'border-box',
  display: 'flex',
  justifyContent: 'center',
  width: '100%',
}))

export const StyledBox = styled(Box)(({ theme }) => ({
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'center',
  width: theme.spacing(256),
}))

export const StyledLink = styled(Link)({
  width: '25%',
  textDecoration: 'none',
})

export const StyledLogo = styled(Box)({
  display: 'flex',
  width: '100%',
  alignItems: 'center',
})

export const StyledTitle = styled(Typography)(({ theme }) => ({
  color: theme.palette.text.secondary,
  paddingLeft: theme.spacing(0.8),
  fontWeight: 600,
}))

export const StyledProductCart = styled(Box)({
  display: 'flex',
})

export const StyledImage = styled('img')(({ theme }) => ({
  width: theme.spacing(2),
  height: theme.spacing(2),
}))
export const StyledProductAmount = styled(Typography)(({ theme }) => ({
  backgroundColor: theme.palette.text.secondary,
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  width: theme.spacing(3),
  height: theme.spacing(3),
  borderRadius: '50%',
  color: theme.palette.background.default,
  marginLeft: theme.spacing(1),
}))

export const StyledSearch = styled(Box)({
  width: '30%',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
})

export const StyledUser = styled(Box)(({ theme }) => ({
  display: 'flex',
  alignItems: 'center',
  color: theme.palette.text.primary,
  padding: theme.spacing(0, 2),
}))

export const StyledLogin = {
  padding: theme.spacing(0, 1),
  fontSize: theme.spacing(1.5),
}

export const StyledLogout = {
  backgroundColor: theme.palette.text.primary,
  color: theme.palette.background.default,
  padding: theme.spacing(0.2, 2),
}

export const StyledUserName = styled(Typography)(({ theme }) => ({
  color: theme.palette.text.primary,
  fontWeight: 600,
  padding: theme.spacing(0, 1),
  textTransform: 'capitalize',
}))

export const StyledUserCart = styled(Box)({
  display: 'flex',
  width: '25%',
  justifyContent: 'flex-end',
  alignItems: 'center',
})

export const StyledSearchIcon = {
  fontSize: theme.spacing(4),
  color: theme.palette.text.secondary,
  paddingLeft: theme.spacing(2),
}
