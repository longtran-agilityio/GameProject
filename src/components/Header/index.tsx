// lib
import LocalMallOutlinedIcon from '@mui/icons-material/LocalMallOutlined'
import AccountCircleSharpIcon from '@mui/icons-material/AccountCircleSharp'
import SearchSharpIcon from '@mui/icons-material/SearchSharp'
import { Typography } from '@mui/material'
import { Link } from 'react-router-dom'

// components
import { StyledButton } from '@webapp/components/Button/Button.module'
import { StyledInputBase } from '@webapp/components/Input/Input.module'

// styles
import {
  StyledHeader,
  StyledLogo,
  StyledTitle,
  StyledImage,
  StyledProductCart,
  StyledProductAmount,
  StyledSearch,
  StyledUser,
  StyledLogin,
  StyledUserName,
  StyledUserCart,
  StyledLogout,
  StyledSearchIcon,
  StyledLink,
  StyledBox,
} from './Header.module'

// context
import { useAuthState, useAuth } from '@webapp/contexts/useAuth'
import { useCart } from '@webapp/contexts/games/cartProvider'

// interface
import { IAuthResponse } from '@webapp/interfaces/user'

// constants
import { PageUrls } from '@webapp/constants/pageUrl'

interface IHeader {
  authenticated: boolean
  onOpen: () => void
}

const Header = ({ authenticated, onOpen }: IHeader) => {
  const { cartsList } = useCart()
  const { logout } = useAuthState()
  const { user } = (useAuth() as IAuthResponse) || {}

  const userName = `${user?.firstName} ${user?.lastName}`
  const handleLogout = () => logout()
  const handleOpen = () => onOpen()

  return (
    <StyledHeader>
      <StyledBox>
        <StyledLink to={PageUrls.HOMEPAGE}>
          <StyledLogo>
            <StyledImage alt='logo' src='/logo.svg' />
            <StyledTitle variant='h3'>Game Store</StyledTitle>
          </StyledLogo>
        </StyledLink>

        <StyledSearch>
          <StyledInputBase placeholder='Search game' type='search' />
          <SearchSharpIcon sx={StyledSearchIcon} />
        </StyledSearch>

        <StyledUserCart>
          {authenticated && (
            <StyledProductCart>
              <StyledButton
                onClick={handleOpen}
                size='large'
                startIcon={<LocalMallOutlinedIcon />}
                disableRipple
              >
                Cart
              </StyledButton>
              <StyledProductAmount>{cartsList?.length}</StyledProductAmount>
            </StyledProductCart>
          )}

          <StyledUser>
            <AccountCircleSharpIcon />
            {authenticated ? (
              <>
                <StyledUserName>{userName}</StyledUserName>
                <StyledButton onClick={handleLogout} variant='contained' sx={StyledLogout}>
                  Logout
                </StyledButton>
              </>
            ) : (
              <>
                <Link to={PageUrls.LOGIN} style={{ textDecoration: 'none' }}>
                  <StyledButton sx={StyledLogin}>LogIn</StyledButton>
                </Link>
                <Typography>|</Typography>
                <Link to={PageUrls.REGISTER} style={{ textDecoration: 'none' }}>
                  <StyledButton sx={StyledLogin}>Register</StyledButton>
                </Link>
              </>
            )}
          </StyledUser>
        </StyledUserCart>
      </StyledBox>
    </StyledHeader>
  )
}

export default Header
