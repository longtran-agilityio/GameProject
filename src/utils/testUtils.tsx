import { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { ThemeProvider } from '@mui/material'
import { theme } from '@webapp/themes/index'

const AllBrowserRouter = ({ children }: { children: React.ReactNode }) => (
  <BrowserRouter>
    <ThemeProvider theme={theme}>{children}</ThemeProvider>
  </BrowserRouter>
)

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllBrowserRouter, ...options })

export * from '@testing-library/react'
export { customRender as render }
