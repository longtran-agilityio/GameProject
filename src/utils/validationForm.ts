export const EmailValidation = {
  required: 'Email is required',
  pattern: {
    value: /^[\w-.]+@([\w-]+\.)+[\w-]{2,4}$/g,
    message: 'Invalid Email',
  },
}

export const PasswordValidation = {
  required: 'Password is required',
  minLength: { value: 8, message: 'min length is 8' },
}

export const FieldEmpty = {
  required: 'This field is required',
}
