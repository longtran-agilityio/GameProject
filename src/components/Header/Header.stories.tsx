// libs
import { ThemeProvider } from '@mui/material'
import { ComponentStory, ComponentMeta } from '@storybook/react'
import { BrowserRouter } from 'react-router-dom'

// component
import Header from '.'
import { theme } from '@webapp/themes/index'

export default {
  title: 'Components/Header',
  component: Header,
} as ComponentMeta<typeof Header>

const Template: ComponentStory<typeof Header> = (args) => (
  <BrowserRouter>
    <ThemeProvider theme={theme}>
      <Header {...args} />
    </ThemeProvider>
  </BrowserRouter>
)

export const HeaderComponent = Template.bind({})
