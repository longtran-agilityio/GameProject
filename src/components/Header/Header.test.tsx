// libs
import { render, screen } from '@webapp/utils/testUtils'

// component
import Header from '.'

const onClick = jest.fn()
const HeaderComponent = () => render(<Header authenticated={true} onOpen={onClick} />)

describe('Header', () => {
  test('should matches snapshot', () => {
    const { container } = HeaderComponent()
    expect(container).toMatchSnapshot()
  })

  test('should get correct value', () => {
    HeaderComponent()
    expect(screen.getByText('Game Store')).toBeInTheDocument()
    expect(screen.getByPlaceholderText('Search game')).toBeInTheDocument()
  })
})
