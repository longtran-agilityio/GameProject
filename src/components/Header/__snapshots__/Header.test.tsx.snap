// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should matches snapshot 1`] = `
<div>
  <header
    class="css-ygfu3e"
  >
    <div
      class="MuiBox-root css-1uynjez"
    >
      <a
        class="css-1tkmtez"
        href="/"
      >
        <div
          class="MuiBox-root css-jjqsud"
        >
          <img
            alt="logo"
            class="css-1t9pz9x"
            src="/logo.svg"
          />
          <h3
            class="MuiTypography-root MuiTypography-h3 css-feyb9i-MuiTypography-root"
          >
            Game Store
          </h3>
        </div>
      </a>
      <div
        class="MuiBox-root css-110k7ip"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-fvh51n-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1tpp6ry-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputTypeSearch css-1g24dm6-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              placeholder="Search game"
              type="search"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-ihdtdm"
              >
                <span
                  class="notranslate"
                >
                  â€‹
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1lgrzs7-MuiSvgIcon-root"
          data-testid="SearchSharpIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </div>
      <div
        class="MuiBox-root css-zbi640"
      >
        <div
          class="MuiBox-root css-gvoll6"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge css-1cjliby-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeLarge css-htszrh-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="LocalMallOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6h-2c0-2.76-2.24-5-5-5S7 3.24 7 6H5c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2zm-7-3c1.66 0 3 1.34 3 3H9c0-1.66 1.34-3 3-3zm7 17H5V8h14v12zm-7-8c-1.66 0-3-1.34-3-3H7c0 2.76 2.24 5 5 5s5-2.24 5-5h-2c0 1.66-1.34 3-3 3z"
                />
              </svg>
            </span>
            Cart
          </button>
          <p
            class="MuiTypography-root MuiTypography-body1 css-j1jdbm-MuiTypography-root"
          />
        </div>
        <div
          class="MuiBox-root css-1n1dh3w"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AccountCircleSharpIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 4c1.93 0 3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5S10.07 6 12 6zm0 14c-2.03 0-4.43-.82-6.14-2.88C7.55 15.8 9.68 15 12 15s4.45.8 6.14 2.12C16.43 19.18 14.03 20 12 20z"
            />
          </svg>
          <p
            class="MuiTypography-root MuiTypography-body1 css-18zlmzq-MuiTypography-root"
          >
            undefined undefined
          </p>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-ri8tu4-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Logout
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </header>
</div>
`;
